web/packages/teleterm/src/services/tshd/cloneableClient.ts

```diff
@@ -0,0 +1,343 @@
+ * Can be produced with `cloneAbortSignal()`.
+ */
+export type CloneableAbortSignal = AbortSignal & {
+  canBePassedThroughContextBridge: true;
+};
+
+/**
+ * Converts a regular `AbortSignal` to the signal
+ * that can be passed over the context bridge.
+ */
+export function cloneAbortSignal(signal: AbortSignal): CloneableAbortSignal {
+  const cloned: Writeable<CloneableAbortSignal> = {
+    canBePassedThroughContextBridge: true,
+    onabort: (...args) => signal.onabort(...args),
+    throwIfAborted: () => signal.throwIfAborted(),
+    reason: signal.reason,
+    aborted: signal.aborted,
+    dispatchEvent: (...args) => signal.dispatchEvent(...args),
+    addEventListener: (type, listener, options) =>
+      signal.addEventListener(type, listener, options),
+    removeEventListener: (type, listener, options) =>
+      signal.removeEventListener(type, listener, options),
+    eventListeners: (...args) => signal.eventListeners(...args),
+    removeAllListeners: (...args) => signal.removeAllListeners(...args),
+  };
+
+  signal.addEventListener(
+    'abort',
+    () => {
+      cloned.reason = signal.reason;
+      cloned.aborted = signal.aborted;
+    },
+    {
+      // Catch the abort event before other listeners to update properties.
+      capture: true,
+      once: true,
+    }
+  );
+
+  return cloned;
+}
+
+type Writeable<T> = {
+  -readonly [P in keyof T]: T[P];
+};
+
+/**
+ * User-provided options for Remote Procedure Calls.
+ *
+ * The only difference from the original `RpcOptions` is the abort signal.
+ * The regular one is replaced with `CloneableAbortSignal`
+ * that can be passed over the context bridge.
+ */
+export type CloneableRpcOptions = Omit<RpcOptions, 'abort'> & {
+  abort?: CloneableAbortSignal;
+};
+
+/**
+ * Describes a client that can be passed over the context bridge.
+ * Errors returned from its methods are converted to `TshdRpcError` objects.
+ */
+export type CloneableClient<Client extends ServiceInfo> = {
+  [Method in keyof Client]: Client[Method] extends (
+    ...args: infer Args
+  ) => infer ReturnType
+    ? (
+        ...args: { [K in keyof Args]: ReplaceRpcOptions<Args[K]> }
+      ) => CloneableCallTypes<ReturnType>
+    : never;
+};
+
+type CloneableCallTypes<T> =
+  T extends UnaryCall<infer Req, infer Res>
+    ? CloneableUnaryCall<Req, Res>
+    : T extends ClientStreamingCall<infer Req, infer Res>
+      ? CloneableClientStreamingCall<Req, Res>
+      : T extends ServerStreamingCall<infer Req, infer Res>
+        ? CloneableServerStreamingCall<Req, Res>
+        : T extends DuplexStreamingCall<infer Req, infer Res>
+          ? CloneableDuplexStreamingCall<Req, Res>
+          : never;
+
+type ReplaceRpcOptions<T> = T extends RpcOptions ? CloneableRpcOptions : T;
+
+/**
+ * A unary RPC call. Can be passed over the context bridge.
+ * Errors are converted to `TshdRpcError` objects.
+ */
+export type CloneableUnaryCall<I extends object, O extends object> = Pick<
```

Comment:
I wanted to say that I need a `CloneableUnaryCall` for types of `reportEvent`, but actually I don't. 
I can do this:
```tsx
  reportUsageEvent: CloneableClient<ITerminalServiceClient>['reportUsageEvent'];
```
instead of this:
```tsx
  reportUsageEvent: (
     input: ReportUsageEventRequest,
     options?: CloneableRpcOptions
   ) => CloneableUnaryCall<ReportUsageEventRequest, apiService.EmptyResponse>;

Reply:
This should be fine for now until we replace `createClient`. In my opinion, when defining types it's better to avoid to referring to types of other "entities".

It's pretty much like Law of Demeter but for types. Such things are not even possible in languages designed with types in mind from the start. 😏